File: js/packages/app/src/__tests__/features/object-references/external/polywrap/object-references-example/src/polywrap/internal/global-functions/invokeGlobalFunction.ts
import { bufferToU32, BaseTypeSerialization, IExternalWrapInstance } from "@polywrap/reim-wrap-js";
import { WrapManifest } from '../../WrapManifest';
import { 
                  
  testExternalGlobalFunction,
   
} from "../../..";
import { TestExternalClassWrapped } from "../../wrapped";

import { TestExternalClass } from "../../..";



export function invoke(buffer: Uint8Array, wrapInstance: IExternalWrapInstance): Promise<Uint8Array> {
  const funcId = bufferToU32(buffer);
  const dataBuffer = buffer.slice(4);

  switch (funcId) {
                                    
    case WrapManifest.Internal.GlobalFunction.TestExternalGlobalFunction:
        return invokeTestExternalGlobalFunctionWrapped(dataBuffer, wrapInstance);
    
    default:
      throw new Error("Unknown function/method: " + funcId.toString());
  }
}

const invokeTestExternalGlobalFunctionWrapped = async (buffer: Uint8Array, wrapInstance: IExternalWrapInstance): Promise<Uint8Array> => {
  const argsBuffer = buffer;

  const args = TestExternalGlobalFunctionArgsWrapped.deserialize(argsBuffer, wrapInstance);

  const result = await testExternalGlobalFunction(
    args.arg,
  );

  
  return BaseTypeSerialization.serialize<string>(result);
};

class TestExternalGlobalFunctionArgs {
  constructor(
    public arg: string,
  ) {
  }
}

class TestExternalGlobalFunctionArgsWrapped {
  constructor(
    public arg: string,
  ) {
  }

  static deserialize(buffer: Uint8Array, wrapInstance: IExternalWrapInstance): TestExternalGlobalFunctionArgs {
    const args = JSON.parse(new TextDecoder().decode(buffer));
  
    return new TestExternalGlobalFunctionArgs(
            
      args.arg,
      
    );
  }  
}

File: rust/app/src/polywrap/external_references/src/polywrap/internal/global-functions/invokeGlobalFunction.rs
use reim_wrap::{ ExternalWrapModule };
use module::{ WrapManifest };
use crate::{ 
                  
  testExternalGlobalFunction,
   
};
use wrapped::{ TestExternalClassWrapped };

use crate::{ TestExternalClass };



async fn invoke(buffer: &[u8], external_module: dyn ExternalWrapModule) -> Vec<u8> {
  let func_id = u32::from_be_bytes(buffer.try_into().expect("Function ID must be 4 bytes"));
  let data_buffer = &buffer[4..];

  match func_id {
                                    
    WrapManifest::Internal::GlobalFunction::TestExternalGlobalFunction =>
        invokeTestExternalGlobalFunctionWrapped(data_buffer, external_module),
    
    _ => panic!("Unknown function/method: ".to_string() + &func_id.to_string()),
  }
}

async fn invokeTestExternalGlobalFunctionWrapped(buffer: &[u8], external_module: dyn ExternalWrapModule) -> Vec<u8> {
  let args_buffer = buffer;

  let args = TestExternalGlobalFunctionArgsWrapped.deserialize(args_buffer, external_module);

  let result = testExternalGlobalFunction(
    args.arg,
  ).await;

  
  BaseTypeSerialization.serialize<string>(result)
};

struct TestExternalGlobalFunctionArgs {
    pub arg: string,
}
 
impl TestExternalGlobalFunctionArgs {
    pub fn new(
        arg: string,
    ) -> Self {
        Self {
            arg,
        }
    }
}

struct TestExternalGlobalFunctionArgsWrapped {
    pub arg: string,
}

impl TestExternalGlobalFunctionArgsWrapped {
    pub fn new(
        arg: string,
    ) -> Self {
        Self {
            arg,
        }
    }

    pub fn deserialize(buffer: &[u8], external_module: dyn ExternalWrapModule) -> TestExternalGlobalFunctionArgs {
        let args: TestExternalGlobalFunctionArgsWrapped = serde_json::from_str(
            str::from_utf8(buffer).expect("Could not convert buffer to string")
        ).expect("JSON was not well-formatted");
       
        TestExternalGlobalFunctionArgs.new(
                        
            args.arg,
            
        )
    }  
}

